File: 09-Intro-to-OOP/3-classes.py
import numpy as np

#Define a class
class Fractional:
    def __init__(self, num, den):
        self.num = num
        self.den = den

    def PrintFraction(self):
        print(self.num, '/', self.den)

    def Plus(self, f2):
        result = Fractional(0,1)
        result.num = self.num*f2.den + f2.num*self.den
        result.den = self.den*f2.den
        return result

    def Minus(self, f2):
        result = Fractional(0,1)
        result.num = self.num*f2.den - f2.num*self.den
        result.den = self.den*f2.den
        return result



f = Fractional(4,3)
f.PrintFraction()

g = Fractional(0,1)
g.num = 5
g.den = 2
g.PrintFraction()

h = Fractional(0,1)
h = f.Plus(g)
h.PrintFraction()

h = f.Minus(g)
h.PrintFraction()

File: 09-Intro-to-OOP/3-classes.cpp
#include<iostream>

//Define the class
class Fractional
{
public:
    int m_num {0};
    int m_den {1};

    void print()
    {
        std::cout << m_num << "/" << m_den << std::endl;
    }

    Fractional operator+(const Fractional & f2);
    Fractional operator-(const Fractional & f2);
    Fractional operator*(const Fractional & f2);
    Fractional operator/(const Fractional & f2);
};

int main(int argc, char **argv)
{
    Fractional f;
    Fractional g {5,6};
    Fractional h;

    f.m_num = 8; // f = 8/1
    h = f + g;
    h.print();
    h = f - g;
    h.print();
    h = f*g;
    h.print();
    h = f/g;
    h.print();

}


Fractional Fractional::operator+(const Fractional & f2)
{
    Fractional result;
    result.m_num = m_num*f2.m_den + f2.m_num*m_den;
    result.m_den = m_den*f2.m_den;
    return result;
}


Fractional Fractional::operator-(const Fractional & f2)
{
    Fractional result;
    result.m_num = m_num*f2.m_den - f2.m_num*m_den;
    result.m_den = m_den*f2.m_den;
    return result;
}

Fractional Fractional::operator*(const Fractional & f2)
{
    Fractional result;
    result.m_num = m_num*f2.m_num;
    result.m_den = m_den*f2.m_den;
    return result;
}

Fractional Fractional::operator/(const Fractional & f2)
{
    Fractional result;
    result.m_num = m_num*f2.m_den;
    result.m_den = m_den*f2.m_num;
    return result;
}
